Fix bugged completion of ~/some in vi mode

JohnoKing opened the pull request:
This pull request fixes the bug reported in att/ast#682. The following sequence fails in vi mode because ksh looks in the wrong part of the `virtual` buffer:

```sh
$ touch ~/testfile
$ ls ~/test<tab>
```

This is fixed by changing `virtual[i]` to `virtual[last_virt]` in the bugged section of code. The other change in this pull request is to make sure listing files in a directory with something like `ls /etc/<tab>` calls the code for Ctrl+L to preserve `ls /etc/` rather than try (and fail) to complete the directory name, producing `ls /etc\n/`. This bugfix was backported from ksh93v- 2013-10-10-alpha. The ksh93v- changelog list a different example for this bug:

>    12-08-31  A bug in vi mode in which ~/tst/make/users/hv<TAB> did not expand correctly has been fixed

~There was also a minor typo in the NEWS file that I fixed as part of this pull request.~

McDutchie commented on NEWS:
```diff
@@ -45,8 +51,8 @@ Any uppercase BUG_* names are modernish shell bug IDs.
-  fixed. Now the real path to the tracked aliases is shown when '-a'
-  is passed to the whence command.
+  fixed. Now the real path to tracked aliases is shown when '-a' is
+  passed to the whence command.
```
You might have noticed that, in `README.md`, I linked the `NEWS` file to GitHub's version of `git blame`, so that people can easily associate news items with commits revealing the full nitty gritty. So I would prefer to have no trivial edits of items that are good enough, as that would confuse this system. If there are serious errors in `NEWS`, they should be fixed in their own commits.

JohnoKing replied:
That makes sense. The typo fix is no longer a part of the pull request.

McDutchie reviewed (commented):
Funny. I can't reproduce this bug on my Mac at all, either on stock 93u+ or on 93u+m. But I can reproduce it on (at least) Linux, FreeBSD, and HP-UX. And on those systems this bug is *not* limited to vi mode, ~~it shows up in emacs mode as well.~~ *it also shows up when both vi mode and emacs mode are disabled.*

And there is another symptom on those systems: *if both vi mode and emacs mode are disabled*, after triggering this bug once, subsequent tab completions produce an extra character at the end of the command line (being the character that was swallowed by the bug) that is impossible to delete. Wow.

Anyway... I'll test this on various systems for a bit before merging this, if you don't mind. We should also figure out a regression test for this bug; it would use pty and go in pty.sh.

Meanwhile I've got one comment on editing the `NEWS` file, see below.

McDutchie commented on src/cmd/ksh93/tests/pty.sh:
```diff
@@ -453,4 +453,21 @@ w fg
 !
 
+# ======
+# Test file name completion in vi mode
```
Please add a dummy `# err_exit #` comment like on the other tests, so that shtests counts it.

JohnoKing committed:
Fix bugged completion of ~/some in vi mode

This commit fixes the bug reported in att/ast#682. The following
sequence fails in vi mode because ksh looks in the wrong part of the
'virtual' buffer:

$ touch ~/testfile
$ ls ~/test<tab>

The fix is to change 'virtual[i]' to 'virtual[last_virt]' in the
bugged section of code. The other changes are to make sure listing
files in a directory with something like 'ls /etc/<tab>' calls the
code for Ctrl+L to preserve 'ls /etc/' rather than try (and fail)
to complete the directory name, producing 'ls /etc\n/'. This bugfix
was backported from ksh93v- 2013-10-10-alpha.

src/cmd/ksh93/edit/vi.c
 - Backport the bugfix from ksh93v- 2013-10-10-alpha for this
   problem.

src/cmd/ksh93/tests/pty.sh
 - Add a regression test for this issue using pty, adjusted slightly
   for a fake home directory in /tmp.
